crates/core/src/utils.rs

```diff
@@ -28,3 +28,22 @@ where
     tokio::task::spawn_local(Abortable::new(future, abort_registration));
     abort_handle
 }
+
+pub trait ResultExt<T, E>: Sized {
+    fn on_error<F: FnOnce(&E)>(self, op: F) -> Result<T, E>;
+    fn on_done<F: FnOnce(&Result<T, E>)>(self, op: F) -> Result<T, E>;
+}
+
+impl<T, E> ResultExt<T, E> for Result<T, E> {
+    fn on_error<F: FnOnce(&E)>(self, op: F) -> Result<T, E> {
+        if let Err(e) = &self {
+            op(e);
+        }
+        self
+    }
+
+    fn on_done<F: FnOnce(&Result<T, E>)>(self, op: F) -> Result<T, E> {
```

Comment:
Is the total more convenient to process than separate success / failure values?

Sadly, `inspect` and `inspect_err` for `Result` are only available in Rust nightly :disappointed:

Reply:
If `inspect`s where available I would go with it, but for now it is more convenient, because you have metric on entry and finish of processing packet, so it is obvious on first glance, that something is wrong.

But I admit I made it this way, because we don't have relay monitoring yet